dockerfile_parse/util.py

```diff
@@ -188,3 +188,113 @@ def remove_nonescaped_quotes(string):
     string = remove_quotes(string)
     string = string.replace('\ ', ' ')
     return string.replace("\\s", "'").replace('\\d', '"')  # restore
+
+
+def split_tuple(text):
+    text_split = text.split('=', 1)
+    if len(text_split) == 2:
+        return tuple(text_split)
+    return None
+
+
+def extract_labels_or_envs(env_replace, envs, instruction_value):
+    shlex_splits_raw = shlex_split(instruction_value,
+                                   env_replace=env_replace, envs=envs)
+
+    key_val_list = []
+    if '=' not in shlex_splits_raw[0]:  # LABEL/ENV name value
+        # split it to first (name) and the rest (value)
+        key_val = instruction_value.split(None, 1)
+        key = strip_quotes(key_val[0])
+        try:
+            val = key_val[1]
+        except IndexError:
+            val = ''
+
+        if env_replace:
+            val = EnvSubst(val, envs).substitute()
+        val = remove_nonescaped_quotes(val)
+
+        key_val_list.append((key, val))
+
+    else:  # LABEL/ENV "name"="value"
+        shlex_splits = {}
+        current_key = None
+        for part in shlex_splits_raw:
+            if current_key:
+                shlex_splits[current_key] = part
+                current_key = None
+            elif part != '=':
+                k_v = split_tuple(part)
+                if k_v:
+                    shlex_splits.update([k_v])
+                else:
+                    current_key = part
```

Comment:
I don't follow the logic here (L221-232), and it's quite different to the code it replaces. What's the intention?

`shlex_splits_raw` is a list of name=value strings at L221. Why are we combining them?